rs/execution_environment/src/hypervisor/tests.rs

```diff
@@ -2182,6 +2182,185 @@ fn ic0_call_cycles_add_has_no_effect_without_ic0_call_perform() {
     );
 }
 
+#[test]
+fn ic0_call_cycles_add128_up_to_deducts_cycles() {
+    let mut test = ExecutionTestBuilder::new()
+        .with_instruction_limit(MAX_NUM_INSTRUCTIONS.get())
+        .build();
+    let wat = r#"
+        (module
+            (import "ic0" "call_new"
+                (func $ic0_call_new
+                    (param i32 i32)
+                    (param $method_name_src i32)    (param $method_name_len i32)
+                    (param $reply_fun i32)          (param $reply_env i32)
+                    (param $reject_fun i32)         (param $reject_env i32)
+                )
+            )
+            (import "ic0" "call_cycles_add128_up_to" (func $ic0_call_cycles_add128_up_to (param i64 i64 i32)))
+            (import "ic0" "call_perform" (func $ic0_call_perform (result i32)))
+            (func (export "canister_update test")
+                (call $ic0_call_new
+                    (i32.const 100) (i32.const 10)  ;; callee canister id = 777
+                    (i32.const 0) (i32.const 18)    ;; refers to "some_remote_method" on the heap
+                    (i32.const 11) (i32.const 22)   ;; fictive on_reply closure
+                    (i32.const 33) (i32.const 44)   ;; fictive on_reject closure
+                )
+                (call $ic0_call_cycles_add128_up_to
+                    (i64.const 0)                       ;; amount of cycles used to be added - high
+                    (i64.const 10000000000)             ;; amount of cycles used to be added - low
+                    (i32.const 200)                     ;; where to write amount of cycles added
+                )
+                (call $ic0_call_perform)
+                drop
+            )
+            (memory 1)
+            (data (i32.const 0) "some_remote_method XYZ")
+            (data (i32.const 100) "\09\03\00\00\00\00\00\00\ff\01")
+        )"#;
+    let initial_cycles = Cycles::new(100_000_000_000);
+    let canister_id = test
+        .canister_from_cycles_and_wat(initial_cycles, wat)
+        .unwrap();
+    let ingress_status = test.ingress_raw(canister_id, "test", vec![]).1;
+    let ingress_state = match ingress_status {
+        IngressStatus::Known { state, .. } => state,
+        IngressStatus::Unknown => unreachable!("Expected known ingress status"),
+    };
+    assert_eq!(IngressState::Processing, ingress_state);
+    assert_eq!(1, test.xnet_messages().len());
+    let mgr = test.cycles_account_manager();
+    let messaging_fee = mgr.xnet_call_performed_fee(test.subnet_size())
+        + mgr.xnet_call_bytes_transmitted_fee(
+            test.xnet_messages()[0].payload_size_bytes(),
+            test.subnet_size(),
+        )
+        + mgr.xnet_call_bytes_transmitted_fee(
+            MAX_INTER_CANISTER_PAYLOAD_IN_BYTES,
+            test.subnet_size(),
+        )
+        + mgr.execution_cost(MAX_NUM_INSTRUCTIONS, test.subnet_size());
+    let transferred_cycles = Cycles::new(10_000_000_000);
```

Comment:
nit: It would probably be nice to confirm that the reply additionally matches the amount requested (i.e. don't rely only on the hardcoded values provided in the test).